crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,413 @@
+        assert_eq!(
+            checkpoint_epoch, node_epoch,
+            "Epoch mismatch after startup recovery. checkpoint epoch: {:?}, node epoch: {:?}",
+            checkpoint_epoch, node_epoch,
+        );
+
+        let next_committee = checkpoint.summary().next_epoch_committee.clone();
+
+        if next_committee.is_some() {
+            self.end_of_epoch = true;
+        }
+
+        let highest_scheduled = checkpoint.sequence_number();
+        let state = self.authority_state.clone();
+        let store = self.checkpoint_store.clone();
+        let metrics = self.metrics.clone();
+
+        pending.push_back(spawn_monitored_task!(async move {
+            while let Err(err) =
+                execute_checkpoint(checkpoint.clone(), state.clone(), store.clone()).await
+            {
+                error!(
+                    "Error while executing checkpoint, will retry in 1s: {:?}",
+                    err
+                );
+                tokio::time::sleep(Duration::from_secs(1)).await;
+                metrics.checkpoint_exec_errors.inc();
+            }
+
+            (checkpoint, next_committee)
+        }));
+
+        self.highest_scheduled_seq_num = Some(highest_scheduled);
+
+        Ok(())
+    }
+}
+
+pub async fn execute_checkpoint(
+    checkpoint: VerifiedCheckpoint,
+    authority_state: Arc<AuthorityState>,
+    checkpoint_store: Arc<CheckpointStore>,
+) -> SuiResult<Vec<SignedTransactionEffects>> {
+    let txes = checkpoint_store
+        .get_checkpoint_contents(&checkpoint.content_digest())?
+        .unwrap_or_else(|| {
+            panic!(
+                "Checkpoint contents for digest {:?} does not exist",
+                checkpoint.content_digest()
+            )
+        })
+        .into_inner();
+
+    execute_transactions(txes, authority_state).await
+}
+
+async fn execute_transactions(
+    transactions: Vec<ExecutionDigests>,
+    authority_state: Arc<AuthorityState>,
+) -> SuiResult<Vec<SignedTransactionEffects>> {
+    let tx_digests: Vec<TransactionDigest> = transactions.iter().map(|tx| tx.transaction).collect();
+    let txns: Vec<VerifiedCertificate> = authority_state
+        .database
+        .epoch_store()
+        .multi_get_pending_certificate(&tx_digests)?
+        .into_iter()
+        .flatten()
+        .collect();
+
+    // TODO once https://github.com/MystenLabs/sui/pull/6157 is landed,
+    // replace with call to `authority_state.add_pending_certs_and_effects`,
+    // which handles reading effects for shared object version numbers
+    // and enqueing transactions with TransactionManager.
+    let digest_to_effects: HashMap<TransactionDigest, TransactionEffects> = authority_state
+        .database
+        .get_effects(&tx_digests)?
+        .into_iter()
+        .map(|fx| {
+            let fx = fx.unwrap();
+            (fx.transaction_digest, fx.data().clone())
+        })
+        .collect();
+
+    for tx in txns.clone() {
```

Comment:
Currently TransactionManager has a very subtle calling convention, for a certificate that may contain shared objects:
1. Set shared locks by order (Narwhal) or from effects (Checkpoint)
2. Add the certificate into the pending_certificates table.
3. Call TransactionManager::enqueue()

(1) must happen before or in the same batch write as (2). Otherwise a node can crash and recover pending certificates without their shared locks set, which violates an invariant.

I do want to simplify the calling convention of TransactionManager, but more cleanups are needed. My plan is to:
1. Add an `enqueue_certificate_with_effects()` function in AuthorityState / TransactionManager`, that runs most of the logic in the current `handle_certificate_with_effects()` without `notify_read_effects()` at the end.
2. Either call `enqueue_certificate_with_effects()` in the current location where certificates from checkpoint are inserted into the `pending_certificates` table,
3. Or buffer certificates from checkpoint into another table, and call `enqueue_certificate_with_effects()` in this module, and cleanup the buffer.

My preference is 1 & 2 but open to suggestions @williampsmith @mystenmark

Reply:
Thanks for calling this out. Chatted with @mwtian about this offline. @bmwill what are your thoughts, as this will affect state sync?

Reply:
I think we'll just need to add another table that state-sync can stick txns in that is epoch-agnostic

Reply:
Yeah this is what I suggested to @mwtian as well. Sounds good

Reply:
i'll work on a PR for that now

Reply:
https://github.com/MystenLabs/sui/pull/6749 for this